src/fwupd_test.py

```diff
@@ -0,0 +1,346 @@
+#!/usr/bin/python3
+# SPDX-License-Identifier: LGPL-2.1+
+
+# to record a string for a device use
+# umockdev-record $PATH
+
+import os
+import subprocess
+import sys
+import tempfile
+import unittest
+import dbusmock
+
+import gi
+from gi.repository import GLib
+from gi.repository import Gio
+
+gi.require_version("UMockdev", "1.0")
+from gi.repository import UMockdev
+
+
+def override_gi_search_path():
+    if "LIBFWUPD_BUILD_DIR" in os.environ:
+        gi.require_version("GIRepository", "2.0")
+        from gi.repository import GIRepository
+
+        GIRepository.Repository.prepend_search_path(
+            os.path.join(os.environ["LIBFWUPD_BUILD_DIR"])
+        )
+        GIRepository.Repository.prepend_library_path(
+            os.path.join(os.environ["LIBFWUPD_BUILD_DIR"])
+        )
+
+
+class FwupdTest(dbusmock.DBusTestCase):
+    DBUS_NAME = "org.freedesktop.fwupd"
+    DBUS_PATH = "/"
+    DBUS_INTERFACE = "org.freedesktop.fwupd"
+
+    @classmethod
+    def setUpClass(cls):
+        if "DAEMON_BUILDDIR" in os.environ:
+            libexecdir = os.environ["DAEMON_BUILDDIR"]
+        else:
+            libexecdir = "@LIBEXECDIR@"
+        cls.daemon_path = os.path.join(libexecdir, "fwupd")
+        for dir in ["STATE_DIRECTORY", "CACHE_DIRECTORY"]:
+            if dir in os.environ:
+                os.makedirs(os.environ[dir], exist_ok=True)
+        os.environ["G_DEBUG"] = "fatal_warnings"
+        GLib.log_set_always_fatal(
+            GLib.LogLevelFlags.LEVEL_WARNING
+            | GLib.LogLevelFlags.LEVEL_ERROR
+            | GLib.LogLevelFlags.LEVEL_CRITICAL
+        )
+        # set up a fake system D-BUS
+        cls.test_bus = Gio.TestDBus.new(Gio.TestDBusFlags.NONE)
+        cls.test_bus.up()
+        try:
+            del os.environ["DBUS_SESSION_BUS_ADDRESS"]
+        except KeyError:
+            pass
+        os.environ["DBUS_SYSTEM_BUS_ADDRESS"] = cls.test_bus.get_bus_address()
```

Comment:
FYI, `self.start_{session,system}_bus()`

Reply:
Thanks!  Will update this tomorrow.